### PR TITLE
Document the `byoei` module

### PR DESCRIPTION
**Pull request by tarekziade**
## Part of https://github.com/elastic/enterprise-search-team/issues/4017

- added docstrings for `byoei.py`
- cleanup dead code

### CODE REVIEW

File: connectors/byoei.py [hunk 1]
```diff
@@ -63,13 +65,14 @@ class Bulker:
     This class runs a coroutine that gets operations out of a `queue` and collects them to
     build and send bulk requests using a `client`
 
-    The bulk requests are controlled in several ways:
+    Arguments:
+
+    - `client` -- an `connectors.es.ESClient` instance
```
Comment (navarone-feekery) [inline connectors/byoei.py:70]:
```suggestion
    - `client` -- a `connectors.es.ESClient` instance
```

File: connectors/byoei.py [hunk 2]
```diff
@@ -167,25 +177,30 @@ async def run(self):
...
     """Grabs data and adds them in the queue for the bulker.
 
-    This class runs a coroutine that puts docs in `queue`, given
-    a document generator.
+    This class runs a coroutine that puts docs in `queue`, given a document generator.
+
+    Arguments:
+    - queue: an `asyncio.Queue` to put docs in
+    - index: the target Elasticsearch index
+    - existing_ids: a list of existing ELasticsearch documents ids found in the index
```
Comment (navarone-feekery) [inline connectors/byoei.py:185]:
```suggestion
    - existing_ids: a list of existing Elasticsearch document ids found in the index
```

**Review by navarone-feekery — COMMENTED** [review]
Added comments on some typos

File: connectors/byoei.py (hunk 1 above)
Comment (artem-shelkovnikov) [reply connectors/byoei.py:70]: What about "an instance of `TYPE`" convention for such cases?

Comment (navarone-feekery) [reply connectors/byoei.py:70]: I like that more ☝🏻

Comment (navarone-feekery) [reply connectors/byoei.py:70]: Updated my comment to reflect that

File: connectors/byoei.py [hunk 3]
```diff
@@ -63,13 +65,14 @@ class Bulker:
     This class runs a coroutine that gets operations out of a `queue` and collects them to
     build and send bulk requests using a `client`
 
-    The bulk requests are controlled in several ways:
+    Arguments:
+
+    - `client` -- an `connectors.es.ESClient` instance
+    - `queue` -- an `asyncio,Queue` instance to pull docs from
```
Comment (navarone-feekery) [inline connectors/byoei.py:71]: Similar to the other comment;
```suggestion
    - `queue` -- an instance of `asyncio,Queue` to pull docs from
```

File: connectors/byoei.py [hunk 4]
```diff
@@ -167,25 +177,30 @@ async def run(self):
...
 
-    This class runs a coroutine that puts docs in `queue`, given
-    a document generator.
+    This class runs a coroutine that puts docs in `queue`, given a document generator.
+
+    Arguments:
+    - queue: an `asyncio.Queue` to put docs in
+    - index: the target Elasticsearch index
+    - existing_ids: a list of existing ELasticsearch documents ids found in the index
+    - filter: a `Filter` instance -- default: `None`
```
Comment (navarone-feekery) [inline connectors/byoei.py:186]: Similar to other comment;
```suggestion
    - filter: an instance of `Filter` -- default: `None`
```

File: connectors/byoei.py [hunk 5]
```diff
@@ -418,6 +435,16 @@ async def async_bulk(
         sync_rules_enabled=False,
         options=None,
     ):
+        """Performs a batch of `_bulk` calls, given a generator of documents
+
+        Arguments:
+        - index: target index
+        - generator: documents generator
+        - pipeline: ingest pipeline settings to pass to the bulk API
+        - filter: a `Filter` instances -- default: `None`
```
Comment (navarone-feekery) [inline connectors/byoei.py:444]:
```suggestion
        - filter: an instance of `Filter` -- default: `None`
```

File: connectors/byoei.py [hunk 6]
```diff
@@ -133,6 +136,13 @@ async def _batch_bulk(self, operations):
         return res
 
     async def run(self):
+        """Creates batches of bulk calls given a queue of documents.
+
+        Exits when the document is "END_DOCS" or "FETCH_ERROR".
```
Comment (timgrein) [inline connectors/byoei.py:141]:
```suggestion
        Exits when the operation is either "END_DOCS" or "FETCH_ERROR".
```

Shouldn't that be "operation"?

File: connectors/byoei.py [hunk 7]
```diff
@@ -133,6 +136,13 @@ async def _batch_bulk(self, operations):
...
+
+        Bulk calls are done concurrently with a maximum number of concurrent
```
Comment (timgrein) [inline connectors/byoei.py:143]:
```suggestion
        Bulk calls are executed concurrently with a maximum number of concurrent
```

File: connectors/byoei.py [hunk 8]
```diff
@@ -234,7 +246,12 @@ async def _deferred_index(self, lazy_download, doc_id, doc, operation):
             }
         )
 
-    async def get_docs(self, generator):
+    async def run(self, generator):
+        """Iterate on a generator of documents to fill a queue of bulk operations for the `Bulker` to consume.
+
+        A document might be discarded if its timestamp has not changed.
+        For documents that contain files to extract, the extraction happens in a separate task.
```
Comment (timgrein) [inline connectors/byoei.py:253]:
```suggestion
        Extraction happens in a separate task, when a document contains files.
```

**Review by timgrein — COMMENTED** [review]
Some small comments and one question.

File: connectors/byoei.py [hunk 9]
```diff
@@ -63,13 +65,14 @@ class Bulker:
...
 
-    The bulk requests are controlled in several ways:
+    Arguments:
+
+    - `client` -- an `connectors.es.ESClient` instance
+    - `queue` -- an `asyncio,Queue` instance to pull docs from
     - `chunk_size` -- a maximum number of operations to send per request
+    - `pipeline` -- ingest pipeline settings to pass to the bulk API
     - `chunk_mem_size` -- a maximum size in MiB for each bulk request
     - `max_concurrency` -- a maximum number of concurrent bulk requests
```
Comment (wangch079) [inline connectors/byoei.py:75]: Might not relate to this PR. But I think we should have default values for fields like `chunk_size`, `chunk_mem_size` and `max_concurrency`.

File: connectors/byoei.py (hunk 4 above)
Comment (wangch079) [reply connectors/byoei.py:186]:
```suggestion
    - filter_: an instance of `Filter` to apply on the fetched document -- default: `None`
```

File: connectors/byoei.py [hunk 10]
```diff
@@ -167,25 +177,30 @@ async def run(self):
...
-    This class runs a coroutine that puts docs in `queue`, given
-    a document generator.
+    This class runs a coroutine that puts docs in `queue`, given a document generator.
+
+    Arguments:
+    - queue: an `asyncio.Queue` to put docs in
+    - index: the target Elasticsearch index
+    - existing_ids: a list of existing ELasticsearch documents ids found in the index
+    - filter: a `Filter` instance -- default: `None`
+    - sync_rules_enabled: if `True`, we apply rules -- default: `False`
```
Comment (wangch079) [inline connectors/byoei.py:187]: Why is this param is added in the first place? I think it's redundant to `Fetcher`, which should only care about `filter_`.

File: connectors/byoei.py (hunk 5 above)
Comment (wangch079) [reply connectors/byoei.py:444]:
```suggestion
        - filter_: an instance of `Filter` to apply on the fetched document  -- default: `None`
```

File: connectors/byoei.py (hunk 6 above)
Comment (tarekziade) [reply connectors/byoei.py:141]: well, technically the queue holds docs, except for that very specific item that is used as a signal.
Maybe we could use the word `item` instead, and say that items in the queue can be documents or on of those constants

File: connectors/byoei.py (hunk 9 above)
Comment (tarekziade) [reply connectors/byoei.py:75]: we do have defaults, they are set in the function that creates the instance, see https://github.com/elastic/connectors-python/blob/main/connectors/byoei.py#L425-L433

but maybe they could be moved at the signature level yeah. good follow-up

File: connectors/byoei.py (hunk 10 above)
Comment (tarekziade) [reply connectors/byoei.py:187]: it's used to instantiate `BasicRuleEngine`, which will run for each document with `self.basic_rule_engine.should_ingest(doc)` to decide if we drop it or not

**Review by artem-shelkovnikov — APPROVED** [review]

**Review by artem-shelkovnikov — APPROVED** [review]

**Pull request merged**